File: examples/retry-backoff/rest/nodejs/retry-backoff.mjs
// Retries a Browserless BQL request with exponential backoff on failure.
//
// Run: node retry-backoff.mjs

const TOKEN = 'YOUR_API_TOKEN_HERE';
const BQL_URL = `https://production-sfo.browserless.io/bql?token=${TOKEN}`;

const MAX_RETRIES = 5;
const BASE_DELAY_MS = 1000;

async function fetchWithRetry(url, options, retries = MAX_RETRIES, delay = BASE_DELAY_MS) {
  for (let attempt = 1; attempt <= retries; attempt++) {
    try {
      const response = await fetch(url, options);
      if (response.ok) return response;
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    } catch (err) {
      if (attempt === retries) throw err;
      console.warn(`Attempt ${attempt} failed: ${err.message}. Retrying in ${delay}ms...`);
      await new Promise((r) => setTimeout(r, delay));
      delay *= 2;
    }
  }
}

const response = await fetchWithRetry(BQL_URL, {
  method: 'POST',
  headers: { 'Content-Type': 'application/json' },
  body: JSON.stringify({
    query: `mutation {
      goto(url: "https://scraping-sandbox.netlify.app/dashboard", waitUntil: networkIdle) { status }
      title { title }
    }`,
    variables: {},
  }),
});

const { data } = await response.json();
console.log('Title:', data.title.title);

File: examples/retry-backoff/rest/csharp/RetryBackoff.cs
// Retries a Browserless BQL request with exponential backoff on failure.
//
// Run: dotnet run

using System.Net.Http;
using System.Text;
using System.Text.Json;

string token = "YOUR_API_TOKEN_HERE";
string endpoint = $"https://production-sfo.browserless.io/bql?token={token}";

const int maxRetries = 5;
int delayMs = 1000;

var payload = JsonSerializer.Serialize(new
{
    query = @"mutation {
      goto(url: ""https://scraping-sandbox.netlify.app/dashboard"", waitUntil: networkIdle) { status }
      title { title }
    }",
    variables = new { },
});

using HttpClient httpClient = new HttpClient();

for (int attempt = 1; attempt <= maxRetries; attempt++)
{
    try
    {
        var content = new StringContent(payload, Encoding.UTF8, "application/json");
        var response = await httpClient.PostAsync(endpoint, content);
        response.EnsureSuccessStatusCode();

        string body = await response.Content.ReadAsStringAsync();
        Console.WriteLine($"Success on attempt {attempt}:");
        Console.WriteLine(body);
        return;
    }
    catch (Exception ex)
    {
        if (attempt == maxRetries) throw;
        Console.WriteLine($"Attempt {attempt} failed: {ex.Message}. Retrying in {delayMs}ms...");
        await Task.Delay(delayMs);
        delayMs *= 2;
    }
}
